### PR TITLE
orders: add Filter() method to GetOrdersRequest and force FilteredOrders return type on wrapper functions

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

After testing recent PR #1054. This attempts to standardize response filtering to requests across all wrapper functions as there were a few missing:

- BTCMarkets GetOrderHistory
- BTSE GetOrderHistory
- FTX GetOrderHistory
- FTX GetActiveOrders
- LBank GetActiveOrders
- LBank GetOrderHistory
- OKGroup GetOrderHistory
- OKGroup GetActiveOrders

master
```
go run ./... getorders --asset=futures --exchange='ftx' --pair=xrp-perp --start="2022-09-10 11:39:09" --end="2022-09-10 11:40:09"
{
 "orders": [
  {
   "exchange": "ftx",
   "id": "189009320677",
   "client_order_id": "web-t384J1WY-1665362274101",
   "base_currency": "XRP",
   "quote_currency": "PERP",
   "asset_type": "futures",
   "order_side": "BUY",
   "order_type": "LIMIT",
   "creation_time": "2022-10-10 00:38:04 +0000",
   "status": "OPEN",
   "price": 0.11,
   "amount": 90,
   "open_volume": 90
  }
 ]
}
```

this brancherino:

```
go run ./... getorders --asset=futures --exchange='ftx' --pair=xrp-perp --start="2022-09-10 11:39:09" --end="2022-09-10 11:40:09"
{}
go run ./... getorders --asset=futures --exchange='ftx' --pair=xrp-perp
{
 "orders": [
  {
   "exchange": "ftx",
   "id": "189009320677",
   "client_order_id": "web-t384J1WY-1665362274101",
   "base_currency": "XRP",
   "quote_currency": "PERP",
   "asset_type": "futures",
   "order_side": "BUY",
   "order_type": "LIMIT",
   "creation_time": "2022-10-10 00:38:04 +0000",
   "status": "OPEN",
   "price": 0.11,
   "amount": 90,
   "open_volume": 90
  }
 ]
}
```

~~**Opened in draft for review commentary and design as I skipped over some caveats in some exchanges which I will need to test and implement.**~~

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/btcmarkets/btcmarkets_wrapper.go [hunk 1]
```diff
@@ -880,19 +881,14 @@ func (b *BTCMarkets) GetActiveOrders(ctx context.Context, req *order.GetOrdersRe
...
-	return resp, nil
+	return req.Filter(b.Name, resp), nil
 }
 
 // GetOrderHistory retrieves account order information
 // Can Limit response to specific order status
-func (b *BTCMarkets) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) ([]order.Detail, error) {
-	if err := req.Validate(); err != nil {
+func (b *BTCMarkets) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) (order.FilteredOrders, error) {
+	err := req.Validate()
```
Comment (gloriousCode) [inline exchanges/btcmarkets/btcmarkets_wrapper.go:890]: Test lacks type and will fail with proper credentials

File: exchanges/btse/btse_wrapper.go [hunk 2]
```diff
@@ -901,8 +895,9 @@ func matchType(input int, required order.Type) bool {
 
 // GetOrderHistory retrieves account order information
 // Can Limit response to specific order status
-func (b *BTSE) GetOrderHistory(ctx context.Context, getOrdersRequest *order.GetOrdersRequest) ([]order.Detail, error) {
-	if err := getOrdersRequest.Validate(); err != nil {
+func (b *BTSE) GetOrderHistory(ctx context.Context, getOrdersRequest *order.GetOrdersRequest) (order.FilteredOrders, error) {
+	err := getOrdersRequest.Validate()
```
Comment (gloriousCode) [inline exchanges/btse/btse_wrapper.go:899]: Test lacks side and will fail with proper credentials

File: exchanges/bybit/bybit_wrapper.go [hunk 3]
```diff
@@ -1557,20 +1558,14 @@ func (by *Bybit) GetActiveOrders(ctx context.Context, req *order.GetOrdersReques
...
-	return orders, nil
+	return req.Filter(by.Name, orders), nil
 }
 
 // GetOrderHistory retrieves account order information
 // Can Limit response to specific order status
-func (by *Bybit) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) ([]order.Detail, error) {
-	if err := req.Validate(); err != nil {
+func (by *Bybit) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) (order.FilteredOrders, error) {
+	err := req.Validate()
```
Comment (gloriousCode) [inline exchanges/bybit/bybit_wrapper.go:1567]: Test lacks side  & type and will fail with proper credentials

File: exchanges/coinbasepro/coinbasepro_wrapper.go [hunk 4]
```diff
@@ -804,45 +807,36 @@ func (c *CoinbasePro) GetActiveOrders(ctx context.Context, req *order.GetOrdersR
...
-	return orders, nil
+	return req.Filter(c.Name, orders), nil
 }
 
 // GetOrderHistory retrieves account order information
 // Can Limit response to specific order status
-func (c *CoinbasePro) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) ([]order.Detail, error) {
-	if err := req.Validate(); err != nil {
+func (c *CoinbasePro) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) (order.FilteredOrders, error) {
+	err := req.Validate()
```
Comment (gloriousCode) [inline exchanges/coinbasepro/coinbasepro_wrapper.go:816]: Test lacks side and will fail with proper credentials

File: exchanges/coinut/coinut_wrapper.go [hunk 5]
```diff
@@ -991,23 +992,18 @@ func (c *COINUT) GetActiveOrders(ctx context.Context, req *order.GetOrdersReques
...
-	return orders, nil
+	return req.Filter(c.Name, orders), nil
 }
 
 // GetOrderHistory retrieves account order information
 // Can Limit response to specific order status
-func (c *COINUT) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) ([]order.Detail, error) {
-	if err := req.Validate(); err != nil {
+func (c *COINUT) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) (order.FilteredOrders, error) {
+	err := req.Validate()
```
Comment (gloriousCode) [inline exchanges/coinut/coinut_wrapper.go:1001]: Test lacks side and will fail with proper credentials

File: exchanges/exmo/exmo_wrapper.go [hunk 6]
```diff
@@ -664,19 +665,14 @@ func (e *EXMO) GetActiveOrders(ctx context.Context, req *order.GetOrdersRequest)
...
-	return orders, nil
+	return req.Filter(e.Name, orders), nil
 }
 
 // GetOrderHistory retrieves account order information
 // Can Limit response to specific order status
-func (e *EXMO) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) ([]order.Detail, error) {
-	if err := req.Validate(); err != nil {
+func (e *EXMO) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) (order.FilteredOrders, error) {
+	err := req.Validate()
```
Comment (gloriousCode) [inline exchanges/exmo/exmo_wrapper.go:674]: Test lacks side and will fail with proper credentials

File: exchanges/ftx/ftx_wrapper.go [hunk 7]
```diff
@@ -1008,13 +1009,14 @@ func (f *FTX) GetActiveOrders(ctx context.Context, getOrdersRequest *order.GetOr
...
-	return resp, nil
+	return getOrdersRequest.Filter(f.Name, resp), nil
 }
 
 // GetOrderHistory retrieves account order information
 // Can Limit response to specific order status
-func (f *FTX) GetOrderHistory(ctx context.Context, request *order.GetOrdersRequest) ([]order.Detail, error) {
-	if err := request.Validate(); err != nil {
+func (f *FTX) GetOrderHistory(ctx context.Context, request *order.GetOrdersRequest) (order.FilteredOrders, error) {
+	err := request.Validate()
```
Comment (gloriousCode) [inline exchanges/ftx/ftx_wrapper.go:1018]: Test lacks side & type and will fail with proper credentials

File: exchanges/gateio/gateio_wrapper.go [hunk 8]
```diff
@@ -816,24 +823,21 @@ func (g *Gateio) GetActiveOrders(ctx context.Context, req *order.GetOrdersReques
...
-	return orders, nil
+	return req.Filter(g.Name, orders), nil
 }
 
 // GetOrderHistory retrieves account order information
 // Can Limit response to specific order status
-func (g *Gateio) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) ([]order.Detail, error) {
-	if err := req.Validate(); err != nil {
+func (g *Gateio) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) (order.FilteredOrders, error) {
+	err := req.Validate()
```
Comment (gloriousCode) [inline exchanges/gateio/gateio_wrapper.go:832]: Test lacks side and will fail with proper credentials

File: exchanges/hitbtc/hitbtc_wrapper.go [hunk 9]
```diff
@@ -769,19 +771,14 @@ func (h *HitBTC) GetActiveOrders(ctx context.Context, req *order.GetOrdersReques
...
-	return orders, nil
+	return req.Filter(h.Name, orders), nil
 }
 
 // GetOrderHistory retrieves account order information
 // Can Limit response to specific order status
-func (h *HitBTC) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) ([]order.Detail, error) {
-	if err := req.Validate(); err != nil {
+func (h *HitBTC) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) (order.FilteredOrders, error) {
+	err := req.Validate()
```
Comment (gloriousCode) [inline exchanges/hitbtc/hitbtc_wrapper.go:780]: Test lacks side and will fail with proper credentials

File: exchanges/huobi/huobi_wrapper.go [hunk 10]
```diff
@@ -1544,19 +1545,14 @@ func (h *HUOBI) GetActiveOrders(ctx context.Context, req *order.GetOrdersRequest
...
-	return orders, nil
+	return req.Filter(h.Name, orders), nil
 }
 
 // GetOrderHistory retrieves account order information
 // Can Limit response to specific order status
-func (h *HUOBI) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) ([]order.Detail, error) {
-	if err := req.Validate(); err != nil {
+func (h *HUOBI) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) (order.FilteredOrders, error) {
+	err := req.Validate()
```
Comment (gloriousCode) [inline exchanges/huobi/huobi_wrapper.go:1554]: Test lacks side and will fail with proper credentials

File: exchanges/itbit/itbit_wrapper.go [hunk 11]
```diff
@@ -583,20 +584,14 @@ func (i *ItBit) GetActiveOrders(ctx context.Context, req *order.GetOrdersRequest
...
-	return orders, nil
+	return req.Filter(i.Name, orders), nil
 }
 
 // GetOrderHistory retrieves account order information
 // Can Limit response to specific order status
-func (i *ItBit) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) ([]order.Detail, error) {
-	if err := req.Validate(); err != nil {
+func (i *ItBit) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) (order.FilteredOrders, error) {
+	err := req.Validate()
```
Comment (gloriousCode) [inline exchanges/itbit/itbit_wrapper.go:593]: Test lacks side and will fail with proper credentials

File: exchanges/kraken/kraken_wrapper.go [hunk 12]
```diff
@@ -1217,19 +1218,14 @@ func (k *Kraken) GetActiveOrders(ctx context.Context, req *order.GetOrdersReques
...
-	return orders, nil
+	return req.Filter(k.Name, orders), nil
 }
 
 // GetOrderHistory retrieves account order information
 // Can Limit response to specific order status
-func (k *Kraken) GetOrderHistory(ctx context.Context, getOrdersRequest *order.GetOrdersRequest) ([]order.Detail, error) {
-	if err := getOrdersRequest.Validate(); err != nil {
+func (k *Kraken) GetOrderHistory(ctx context.Context, getOrdersRequest *order.GetOrdersRequest) (order.FilteredOrders, error) {
+	err := getOrdersRequest.Validate()
```
Comment (gloriousCode) [inline exchanges/kraken/kraken_wrapper.go:1227]: Test lacks side and will fail with proper credentials

File: exchanges/lbank/lbank_wrapper.go [hunk 13]
```diff
@@ -713,13 +714,14 @@ func (l *Lbank) GetActiveOrders(ctx context.Context, getOrdersRequest *order.Get
...
-	return finalResp, nil
+	return getOrdersRequest.Filter(l.Name, finalResp), nil
 }
 
 // GetOrderHistory retrieves account order information *
 // Can Limit response to specific order status
-func (l *Lbank) GetOrderHistory(ctx context.Context, getOrdersRequest *order.GetOrdersRequest) ([]order.Detail, error) {
-	if err := getOrdersRequest.Validate(); err != nil {
+func (l *Lbank) GetOrderHistory(ctx context.Context, getOrdersRequest *order.GetOrdersRequest) (order.FilteredOrders, error) {
+	err := getOrdersRequest.Validate()
```
Comment (gloriousCode) [inline exchanges/lbank/lbank_wrapper.go:723]: Test lacks type and will fail with proper credentials

File: exchanges/lbank/lbank_wrapper.go [hunk 14]
```diff
@@ -655,8 +655,9 @@ func (l *Lbank) WithdrawFiatFundsToInternationalBank(_ context.Context, _ *withd
 }
 
 // GetActiveOrders retrieves any orders that are active/open
-func (l *Lbank) GetActiveOrders(ctx context.Context, getOrdersRequest *order.GetOrdersRequest) ([]order.Detail, error) {
-	if err := getOrdersRequest.Validate(); err != nil {
+func (l *Lbank) GetActiveOrders(ctx context.Context, getOrdersRequest *order.GetOrdersRequest) (order.FilteredOrders, error) {
+	err := getOrdersRequest.Validate()
```
Comment (gloriousCode) [inline exchanges/lbank/lbank_wrapper.go:659]: Could you add a test for this?

File: exchanges/yobit/yobit_wrapper.go [hunk 15]
```diff
@@ -617,36 +617,33 @@ func (y *Yobit) GetActiveOrders(ctx context.Context, req *order.GetOrdersRequest
...
-	return orders, nil
+	return req.Filter(y.Name, orders), nil
 }
 
 // GetOrderHistory retrieves account order information
 // Can Limit response to specific order status
-func (y *Yobit) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) ([]order.Detail, error) {
-	if err := req.Validate(); err != nil {
+func (y *Yobit) GetOrderHistory(ctx context.Context, req *order.GetOrdersRequest) (order.FilteredOrders, error) {
+	err := req.Validate()
```
Comment (gloriousCode) [inline exchanges/yobit/yobit_wrapper.go:626]: Test lacks side and will fail with proper credentials

File: exchanges/ftx/ftx_wrapper.go [hunk 16]
```diff
@@ -918,8 +918,9 @@ func (f *FTX) GetWebsocket() (*stream.Websocket, error) {
 }
 
 // GetActiveOrders retrieves any orders that are active/open
-func (f *FTX) GetActiveOrders(ctx context.Context, getOrdersRequest *order.GetOrdersRequest) ([]order.Detail, error) {
-	if err := getOrdersRequest.Validate(); err != nil {
+func (f *FTX) GetActiveOrders(ctx context.Context, getOrdersRequest *order.GetOrdersRequest) (order.FilteredOrders, error) {
+	err := getOrdersRequest.Validate()
```
Comment (gloriousCode) [inline exchanges/ftx/ftx_wrapper.go:922]: Anyside and anytype pass validation, however, they do not translate well to actually make a request. We should not populate the FTX request with those fields when any

File: exchanges/order/order_test.go [hunk 17]
```diff
@@ -1869,3 +1887,50 @@ func TestDeriveCancel(t *testing.T) {
...
+	request.EndTime = time.Unix(1336, 0)
+	request.StartTime = time.Unix(1337, 0)
+
+	shinyAndClean = request.Filter("test", orders)
+
+	if len(shinyAndClean) != 8 {
+		t.Fatalf("received: '%v' but expected: '%v'", len(shinyAndClean), 8)
+	}
+
+	if !shinyAndClean[0].Pair.Equal(btcltc) {
```
Comment (gloriousCode) [inline exchanges/order/order_test.go:1933]: I think we could make this a little better. Perhaps a `for` to ensure all returned items are filtered appropriately. Or perhaps orderID checks

File: exchanges/order/orders.go [hunk 18]
```diff
@@ -1106,28 +1107,52 @@ func (c *Cancel) Validate(opt ...validate.Checker) error {
...
 	}
 	return nil
 }
 
+// Filter reduces slice by optional fields
+func (g *GetOrdersRequest) Filter(exch string, orders []Detail) FilteredOrders {
+	filtered := make([]Detail, len(orders))
+	copy(filtered, orders)
+	FilterOrdersByPairs(&filtered, g.Pairs)
+	FilterOrdersByType(&filtered, g.Type)
```
Comment (gloriousCode) [inline exchanges/order/orders.go:1147]: Having all of these requests are likely to have consequeces. See the implementation of this function with the following:

```go
if AnyType == orderType || len(*orders) == 0 {
		return
	}

	target := 0
	for i := range *orders {
		if (*orders)[i].Type == orderType {
			(*orders)[target] = (*orders)[i]
			target++
		}
	}
```
If you provide anything other than `ANY`, then the filter _must_ match, even if its `unknowntype` and therefore, we will filter any orders that do not populate that data in their API response. My next review will focus on the concequences. A simple fix would be to allow `FilterOrdersByType` and `FilterOrdersBySide` to always accept unpopulated fields, or for `GetOrdersRequest` to have an additional field which will allow non-populated fields to be returned

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Nice! This is just a quick review of the easy nits.

File: exchanges/bybit/bybit_test.go [hunk 19]
```diff
@@ -2794,6 +2794,8 @@ func TestGetActiveOrders(t *testing.T) {
 	var getOrdersRequestSpot = order.GetOrdersRequest{
 		Pairs:     currency.Pairs{pair},
 		AssetType: asset.Spot,
+		Side:      order.AnySide,
+		Type:      order.AnyType,
 	}
 
 	_, err = b.GetActiveOrders(context.Background(), &getOrdersRequestSpot)
```
Comment (gloriousCode) [inline exchanges/bybit/bybit_test.go:2801]: Please update the remaining `order.GetOrdersRequest` scenarios to have `Side` and `Type`

File: exchanges/ftx/ftx_test.go [hunk 20]
```diff
@@ -1099,6 +1099,8 @@ func TestGetActiveOrders(t *testing.T) {
 	cp := currency.NewPairWithDelimiter(currency.BTC.String(), currency.USDT.String(), "/")
 	orderReq.Pairs = append(orderReq.Pairs, cp)
 	orderReq.AssetType = asset.Spot
+	orderReq.Side = order.AnySide
+	orderReq.Type = order.AnyType
 	_, err := f.GetActiveOrders(context.Background(), &orderReq)
```
Comment (gloriousCode) [inline exchanges/ftx/ftx_test.go:1104]: I love the use of the validator checks. Could you please add another scenario for `ImmediateOrCancel` to test the expected error?

File: exchanges/ftx/ftx_test.go [hunk 21]
```diff
@@ -1114,6 +1116,8 @@ func TestGetOrderHistory(t *testing.T) {
 	cp := currency.NewPairWithDelimiter(currency.BTC.String(), currency.USDT.String(), "/")
 	orderReq.Pairs = append(orderReq.Pairs, cp)
 	orderReq.AssetType = asset.Spot
+	orderReq.Side = order.AnySide
+	orderReq.Type = order.AnyType
 	_, err := f.GetOrderHistory(context.Background(), &orderReq)
```
Comment (gloriousCode) [inline exchanges/ftx/ftx_test.go:1121]: I love the use of the validator checks. Could you please add another scenario for `ImmediateOrCancel` to test the expected error? Along with a side of `MissingData` or some other invalid option

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Thanks for making all those changes! Couple more things, but I'm happy with how you've handled it all

**Review by gloriousCode — APPROVED** [review]
tACK! Thanks for making those changes. Verified filtering via rpc as well

**Review by thrasher- — APPROVED** [review]
tACK, a more comprehensive solution which catches the previous gotcha!

**Pull request merged**